src/config/field-overrides/relationshipType.js

```diff
@@ -156,12 +186,44 @@ class Constraint extends Component {
                         error: e,
                     });
                 });
+        } else {
+            this.setState({ loading: false });
+        }
+    }
+
+    componentDidUpdate(prevProps) {
+        const prevEntity = this.getSelectedRelationshipEntity(prevProps);
+        const entity = this.getSelectedRelationshipEntity();
+        if (prevEntity !== entity) {
+            const optionsLoading = this.initalizeOptionsLoadingState();
+            this.setState({ optionsLoading });
         }
-        this.setState({ loading: false });
     }
 
-    getSelectedRelationshipEntity = () => {
-        const props = this.props;
+    initalizeOptionsLoadingState = () => {
+        const entity = this.getSelectedRelationshipEntity();
+        const modelTypes = modelTypesForRelationshipEntity[entity];
+        // initialize loading state for modelTypes
+        const optionsLoadingState = modelTypes.reduce((acc, modelOpts) => {
+            const modelType = modelOpts.modelType;
+            acc[modelType] = this.shouldRenderModelType(modelType);
+            return acc;
+        }, {});
+        return optionsLoadingState;
+    };
+
+    isLoading = () => {
+        const optionsLoading = this.isOptionsLoading();
+        return this.state.loading || optionsLoading;
+    };
```

Comment:
```suggestion
    isLoading = () => {
        return this.state.loading || this.isOptionsLoading();
    };
```
This is a line less code and will also prevent `this.isOptionsLoading()` from executing.